Add netlink multicast test

jsimonetti opened the pull request:
Add a proper CI test for netlink groups.

mdlayher commented on conn_linux_integration_test.go:
```diff
@@ -0,0 +1,67 @@
+	c, err := Dial(0, cfg) // dials NETLINK_ROUTE
+	if err != nil {
+		t.Fatalf("failed to dial: %v", err)
+
```
Yank this empty line.

mdlayher commented on conn_linux_integration_test.go:
```diff
@@ -0,0 +1,67 @@
+
+	// routine for receiving any messages
+	recv := func() {
+		data, er := c.Receive()
```
Let's stick with `err` for consistency.

mdlayher commented on conn_linux_integration_test.go:
```diff
@@ -0,0 +1,67 @@
+	recv := func() {
+		data, er := c.Receive()
+		if er != nil {
+			fmt.Printf("error in receive: %s\n", err)
```
`t.Fatalf`, or maybe even `panic(fmt.Sprintf())`.  Not sure what happens when you `t.Fatal` in a goroutine.

mdlayher replied:
Don't need newlines when using `t.Fatal`.

mdlayher commented on conn_linux_integration_test.go:
```diff
@@ -0,0 +1,67 @@
+	case got = <-in:
+		break
+	case <-timeout:
+		t.Fatal("did not receive any messages after 5 seconds\n")
```
Don't need newlines when using `t.Fatal`.

mdlayher commented on conn_linux_integration_test.go:
```diff
@@ -0,0 +1,67 @@
+	cmd := exec.Command("sudo", "ip", "tuntap", "add", ifName, "mode", "tun")
+	err := cmd.Start()
+	if err != nil {
+		t.Fatalf("error creating tuntap device: %s\n", err)
```
Don't need newlines when using `t.Fatal`.

mdlayher commented on conn_linux_integration_test.go:
```diff
@@ -0,0 +1,67 @@
+		cmd := exec.Command("sudo", "ip", "link", "del", ifName)
+		err := cmd.Start()
+		if err != nil {
+			t.Fatalf("error removing tuntap device: %s\n", err)
```
Don't need newlines when using `t.Fatal`.

mdlayher commented on conn_linux_integration_test.go:
```diff
@@ -0,0 +1,67 @@
+		if err != nil {
+			t.Fatalf("error removing tuntap device: %s\n", err)
+		}
+		cmd.Wait()
```
Might as well just check this error too.

mdlayher commented on conn_linux_integration_test.go:
```diff
@@ -0,0 +1,67 @@
+		t.Fatal("did not receive any messages after 5 seconds\n")
+		break
+	}
+	t.Logf("received netlink data: %#v\n", got)
```
Obviously this isn't a `rtnetlink` package, but are there any kind of header values we can inspect quickly here just to determine that this is a valid response and not something bogus?

jsimonetti commented:
I have updated the PR with the requested style changes.

I also added a check in the received message.
The message should be a ifinfomsg with the interface name equal to the interface we created during the test.
To go much further, we probably have to decode the message data.

mdlayher commented on conn_linux_integration_test.go:
```diff
@@ -0,0 +1,81 @@
+	copy(got, data[0].Data[20:len(ifName)+20])
+
+	if !reflect.DeepEqual(want, got) {
+		t.Fatalf("received message doesn not mention ifName %s", ifName)
```
*does, and let's go with `%q` so any weird whitespace or whatever is easily spotted if this fails.

mdlayher commented on conn_linux_integration_test.go:
```diff
@@ -0,0 +1,81 @@
+	case data = <-in:
+		break
+	case <-timeout:
+		t.Fatal("did not receive any messages after 5 seconds")
```
`panic()` is nice in the case of timeouts because it'll give a stack trace showing where the goroutines are hung up.

mdlayher commented on conn_linux_integration_test.go:
```diff
@@ -0,0 +1,81 @@
+	"os/exec"
+	"testing"
+	"time"
+
```
Weird, mind removing this?

mdlayher commented on conn_linux_integration_test.go:
```diff
@@ -0,0 +1,81 @@
+		t.Fatalf("error creating tuntap device: %s", err)
+		return func() {}
+	}
+	cmd.Wait()
```
Check the error.

mdlayher commented on conn_linux_integration_test.go:
```diff
@@ -0,0 +1,81 @@
+		if err != nil {
+			panic(fmt.Sprintf("error removing tuntap device: %s", err))
+		}
+		cmd.Wait()
```
Check the error.

jsimonetti commented:
One addition commit to address the requested changes.
Feel free to squash this at merge time btw.
